File: parser/rule/import/import.go
/*
   The Fluent Programming Language
   -----------------------------------------------------
   This code is released under the GNU GPL v3 license.
   For more information, please visit:
   https://www.gnu.org/licenses/gpl-3.0.html
   -----------------------------------------------------
   Copyright (c) 2025 Rodrigo R. & All Fluent Contributors
   This program comes with ABSOLUTELY NO WARRANTY.
   For details type `fluent l`. This is free software,
   and you are welcome to redistribute it under certain
   conditions; type `fluent l -f` for details.
*/

package _import

import (
	"fluent/ast"
	"fluent/parser/error"
	"fluent/token"
)

// ProcessImport processes a list of tokens representing an import statement.
// It expects exactly two tokens: the 'import' keyword and a string literal.
// If the tokens are not as expected, it returns an error.
// If the tokens are valid, it returns an AST node representing the import statement.
//
// Parameters:
//
//	tokens []token.Token - A slice of tokens representing the import statement.
//
// Returns:
//
//	ast.AST - The AST node representing the import statement.
//	error.Error - An error object if the tokens are invalid.
func ProcessImport(tokens []token.Token) (*ast.AST, *error.Error) {
	// The first token is always an import keyword
	// We are not going to process it

	if len(tokens) != 2 {
		return nil, &error.Error{
			Line:     tokens[0].Line,
			Column:   tokens[0].Column,
			File:     &tokens[0].File,
			Expected: []ast.Rule{ast.StringLiteral},
		}
	}

	// Basic validation for the string literal
	if tokens[1].TokenType != token.StringLiteral {
		return nil, &error.Error{
			Line:     tokens[1].Line,
			Column:   tokens[1].Column,
			File:     &tokens[1].File,
			Expected: []ast.Rule{ast.StringLiteral},
		}
	}

	// Directly parse the import rule
	child := ast.AST{
		Rule: ast.Import,
		Children: &[]*ast.AST{
			{
				Rule:     ast.StringLiteral,
				Value:    tokens[1].Value,
				Line:     tokens[1].Line,
				Column:   tokens[1].Column,
				File:     &tokens[1].File,
				Children: &[]*ast.AST{},
			},
		},
		Line:   tokens[0].Line,
		Column: tokens[0].Column,
		File:   &tokens[0].File,
	}

	return &child, nil
}

File: parser/rule/import/import.cpp
/*
        ==== The Zelix Programming Language ====
---------------------------------------------------------
  - This file is part of the Zelix Programming Language
    codebase. Zelix is a fast, statically-typed and
    memory-safe programming language that aims to
    match native speeds while staying highly performant.
---------------------------------------------------------
  - Zelix is categorized as free software; you can
    redistribute it and/or modify it under the terms of
    the GNU General Public License as published by the
    Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.
---------------------------------------------------------
  - Zelix is distributed in the hope that it will
    be useful, but WITHOUT ANY WARRANTY; without even
    the implied warranty of MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE. See the GNU General Public
    License for more details.
---------------------------------------------------------
  - You should have received a copy of the GNU General
    Public License along with Zelix. If not, see
    <https://www.gnu.org/licenses/>.
*/

//
// Created by rodri on 8/16/25.
//

#include "import.h"
#include "parser/expect.h"
#include "parser/parser.h"

using namespace zelix;

void parser::rule::imp(
    ast *&root,
    container::stream<lexer::token *> &tokens,
    const bool &top_level,
    memory::lazy_allocator<ast> &allocator,
    const lexer::token *const &trace
)
{
    // Make sure we are at the top level
    if (!top_level)
    {
        global_err.type = ILLEGAL_IMPORT;
        global_err.column = trace->column;
        global_err.line = trace->line;
        throw except::exception("Illegal import statement outside of top-level scope");
    }

    expect(tokens, lexer::token::STRING_LITERAL);
    const auto path = tokens.next()
        .get();

    // Expect a semicolon
    expect(tokens, lexer::token::SEMICOLON);
    tokens.next(); // Consume the semicolon

    // Create the import node
    ast *import_node = allocator.alloc();
    import_node->rule = ast::IMPORT;
    import_node->value = path->value;
    import_node->line = path->line;
    import_node->column = path->column;
    root->children.push_back(import_node);
}
